R/setkey.R

```diff
@@ -155,16 +155,24 @@ setreordervec = function(x, order) .Call(Creorder, x, order)
 # The others (order, sort.int etc) are turned off to protect ourselves from using them internally, for speed and for
 # consistency; e.g., consistent twiddling of numeric/integer64, NA at the beginning of integer, locale ordering of character vectors.
 
-is.sorted = function(x, by=seq_along(x)) {
+is.sorted = function(x, by=seq_along(x), retOrd=FALSE) {
   if (is.list(x)) {
-    warning("Use 'if (length(o <- forderv(DT,by))) ...' for efficiency in one step, so you have o as well if not sorted.")
+    # for efficient use via retOrd argument see note in ?is.sorted
     # could pass through a flag for forderv to return early on first FALSE. But we don't need that internally
     # since internally we always then need ordering, an it's better in one step. Don't want inefficiency to creep in.
-    # This is only here for user/debugging use to check/test valid keys; e.g. data.table:::is.sorted(DT,by)
-    0L == length(forderv(x,by,retGrp=FALSE,sort=TRUE))
-  } else {
+    o = forderv(x,by,retGrp=FALSE,sort=TRUE)
+    ans = 0L == length(o)
+    if (isTRUE(retOrd))
+      ans = setattr(copy(ans), "order", o)
+    ans
+  } else if (is.null(x)) { # NULL does not satisfy C isVectorAtomic
+    NA
+  } else if (is.atomic(x)) {
+    if (isTRUE(retOrd)) stop("retOrd works only for data.table/list input")
```

Comment:
If I saw this error message, I would make my vector a list. Should we automatically do ```list(x)``` for convenience?

Reply:
On one hand it would be nice, but fsorted that we use for atomic vectors is likely to be much more efficient. So by default it definitely not good to wrap it in list.nooooooooooooooooooooooooooooooooooooooooooooooooo!

Reply:
Only wrap it if ```retOrd == TRUE```.  That arg would seem to trump performance considerations of fsorted.